_posts/2022-05-18-torchopt/torchopt.Rmd

```diff
@@ -0,0 +1,356 @@
+
+[^1]: For example, by creating an issue on GitHub, in either the [blog's repo](https://github.com/rstudio/ai-blog) or that of the respective software (e.g., [keras](https://github.com/rstudio/keras), [torch](https://github.com/mlverse/torch), ...)
+
+## `torchopt`
+
+[`torchopt`](https://github.com/e-sensing/torchopt) is a package developed by [Gilberto Camara](https://gilbertocamara.org/) and colleagues at [National Institute for Space Research, Brazil](https://www.gov.br/inpe/pt-br).
+
+By the look of it, the package's reason of being is rather self-evident. `torch` itself does not -- nor should it -- implement all the newly-published, potentially-useful-for-your-purposes optimization algorithms out there.[^2] The algorithms assembled here, then, are probably exactly those the authors were most eager to experiment with in their own work. As of this writing, they comprise, amongst others, various members of the popular *ADA*\* and \**ADAM*\* families.[^3] And we may safely assume the list will grow over time.
+
+[^2]: For why it shouldn't, and an overview of (some of) the things `torch` users can do, see also our recent post on [extending `torch`](https://blogs.rstudio.com/ai/posts/2022-04-27-torch-outside-the-box/).
+
+[^3]: See the [Readme](https://github.com/e-sensing/torchopt) for an up-to-date listing.
+
+I'm going to introduce the package by highlighting something that technically, is "merely" a utility function, but to the user, can be extremely helpful: the ability to, for an arbitrary optimizer and an arbitrary test function, plot the steps taken in optimization.
+
+While it's true that I have no intent of comparing (let alone analyzing) different strategies, there is one that, to me, stands out in the list: ADAHESSIAN [@abs-2006-00719], a second-order algorithm designed to scale to large neural networks. I'm especially curious to see how it behaves as compared to L-BFGS, the second-order "classic" available from base `torch` we've had a [dedicated blog post](https://blogs.rstudio.com/ai/posts/2021-04-22-torch-for-optimization/) about last year.
+
+## The way it works
+
+The utility function in question is named `test_optim()`. The only required argument concerns the optimizer to try (`optim`). But you'll likely want to tweak three others as well:
+
+-   `test_fn`: To use a test function different from the default (`beale`). You can choose among the many provided in `torchopt`, or you can pass in your own. In the latter case, you also need to provide information about search domain and starting points. (We'll see that in an instant.)
+-   `steps`: To set the number of optimization steps.
+-   `opt_hparams`: To modify optimizer hyperparameters; most notably, the learning rate.
+
```

Comment:
Maybe mention it can also produce the GIF's?

Reply:
honestly, my experiences with those weren't so good (wild flickering, etc.)
I think I'll forego that :-)